### PR TITLE
Move in message marshalling functions from 'envelopespec' package.

### PR DESCRIPTION
**Pull request by danilvpetrov**
#### What change does this introduce?

This PR moves the message marshalling functions from [`dogmatiq/envelopespec`](https://github.com/dogmatiq/envelopspec) package into this repository.  

#### What issues does this relate to?

Fixes https://github.com/dogmatiq/marshalkit/issues/40.

#### Why make this change?

This change is required to relocate message marshalling functions to the current repository from [`dogmatiq/envelopespec`](https://github.com/dogmatiq/envelopspec) package as they are not related to the envelope spec definitions, but to message marshalling. 

#### Is there anything you are unsure about?

No.

### CODE REVIEW

**Comment by danilvpetrov** [comment]
I had to slightly modify the imports in `envelope_test.go` file compared to what it was before in the [`dogmatiq/envelopespec`](https://github.com/dogmatiq/envelopespec) package as there is a conflict with the `marshalkit.Marshaler` type and `marhalkit/fixtures.Marshaler` variable in the tests.

**Comment by jmalloc** [comment]
Should we wait until `envelopespec` has been moved/merged/tagged in `interopspec` before merging this?

**Comment by danilvpetrov** [comment]
Yeah we probably should as the imports and refs would have to be changed.

File: envelope.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// within envelopes.
+func MustMarshalEnvelopeTime(t time.Time) string {
+	if t.IsZero() {
+		return ""
+	}
+
+	return t.Format(time.RFC3339Nano)
+}
+
+// UnmarshalEnvelopeTime unmarshals a t from its RFC-3339 representation, as
```
Comment (koden-km) [inline envelope.go:69]:
```suggestion
// UnmarshalEnvelopeTime unmarshals t from its RFC-3339 representation, as
```

File: envelope.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"time"
+
+	"github.com/dogmatiq/configkit"
+	"github.com/dogmatiq/dogma"
+	"github.com/dogmatiq/interopspec/envelopespec"
+)
+
+// MustMarshalMessageIntoEnvelope marshals a Dogma message into an envelope.
```
Comment (jmalloc) [inline envelope.go:11]:
```suggestion
// MustMarshalMessageIntoEnvelope marshals a Dogma message into an envelopespec.Envelope.
```

File: envelope.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		// own requirements on the format of the media-type.
+		panic(err)
+	}
+
+	env.PortableName = n
+	env.MediaType = p.MediaType
+	env.Data = p.Data
+}
+
+// UnmarshalMessageFromEnvelope unmarshals a Dogma message from an envelope.
```
Comment (jmalloc) [inline envelope.go:31]:
```suggestion
// UnmarshalMessageFromEnvelope unmarshals a Dogma message from an envelopespec.Envelope
```

File: envelope.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	p := Packet{
+		MediaType: env.MediaType,
+		Data:      env.Data,
+	}
+
+	return vm.Unmarshal(p)
+}
+
+// MustMarshalEnvelopeIdentity marshals id to its protocol buffers
+// representation, as used within message envelopes.
```
Comment (jmalloc) [inline envelope.go:45]:
```suggestion
// representation, as used within envelopespec.Envelope.
```

File: envelope.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func MustMarshalEnvelopeIdentity(id configkit.Identity) *envelopespec.Identity {
+	return &envelopespec.Identity{
+		Name: id.Name,
+		Key:  id.Key,
+	}
+}
+
+// UnmarshalEnvelopeIdentity unmarshals id from its protocol buffers
+// representation, as used within message envelopes.
```
Comment (jmalloc) [inline envelope.go:54]:
```suggestion
// representation, as used within envelopespec.Envelope.
```

File: envelope.go (hunk 1 above)
Comment (danilvpetrov) [inline envelope.go:69]:
```suggestion
// UnmarshalEnvelopeTime unmarshals t from its RFC-3339 representation, as
```

File: envelope_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	. "github.com/dogmatiq/dogma/fixtures"
+	. "github.com/dogmatiq/interopspec/envelopespec"
+	. "github.com/dogmatiq/marshalkit"
+	"github.com/dogmatiq/marshalkit/fixtures"
+	. "github.com/jmalloc/gomegax"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("func MarshalMessage()", func() {
```
Comment (koden-km) [inline envelope_test.go:16]:
```suggestion
var _ = Describe("func MustMarshalMessageIntoEnvelope()", func() {
```

File: envelope_test.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+			&Envelope{
+				PortableName: fixtures.MessageAPortableName,
+				MediaType:    fixtures.MessageA1Packet.MediaType,
+				Data:         fixtures.MessageA1Packet.Data,
+			},
+		))
+	})
+})
+
+var _ = Describe("func UnmarshalMessage()", func() {
```
Comment (koden-km) [inline envelope_test.go:36]:
```suggestion
var _ = Describe("func UnmarshalMessageFromEnvelope()", func() {
```

File: envelope.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+func UnmarshalEnvelopeIdentity(id *envelopespec.Identity) (configkit.Identity, error) {
+	return configkit.NewIdentity(id.Name, id.Key)
+}
+
+// MustMarshalEnvelopeTime marshals t to its RFC-3339 representation, as used
+// within envelopes.
```
Comment (jmalloc) [inline envelope.go:60]:
```suggestion
// within envelopespec.Envelope.
```

File: envelope_test.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+			Data:         fixtures.MessageA1Packet.Data,
+		}
+
+		m, err := UnmarshalMessageFromEnvelope(fixtures.Marshaler, env)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(m).To(Equal(MessageA1))
+	})
+})
+
+var _ = Describe("func MarshalIdentity()", func() {
```
Comment (koden-km) [inline envelope_test.go:50]:
```suggestion
var _ = Describe("func MustMarshalEnvelopeIdentity()", func() {
```

File: envelope.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+func MustMarshalEnvelopeTime(t time.Time) string {
+	if t.IsZero() {
+		return ""
+	}
+
+	return t.Format(time.RFC3339Nano)
+}
+
+// UnmarshalEnvelopeTime unmarshals a t from its RFC-3339 representation, as
+// used within envelopes.
```
Comment (jmalloc) [inline envelope.go:70]:
```suggestion
// used within envelopespec.Envelope.
```

File: envelope_test.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+		Expect(out).To(EqualX(
+			&Identity{
+				Name: "<name>",
+				Key:  "<key>",
+			},
+		))
+	})
+})
+
+var _ = Describe("func UnmarshalIdentity()", func() {
```
Comment (koden-km) [inline envelope_test.go:64]:
```suggestion
var _ = Describe("func UnmarshalEnvelopeIdentity()", func() {
```

File: envelope_test.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+		out, err := UnmarshalEnvelopeIdentity(in)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(out).To(Equal(
+			configkit.MustNewIdentity("<name>", "<key>"),
+		))
+	})
+})
+
+var _ = Describe("func MarshalTime()", func() {
```
Comment (koden-km) [inline envelope_test.go:79]:
```suggestion
var _ = Describe("func MustMarshalEnvelopeTime()", func() {
```

File: envelope_test.go [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+	It("returns an empty string if the time is the zero-value", func() {
+		var in time.Time
+
+		out := MustMarshalEnvelopeTime(in)
+		Expect(out).To(Equal(""))
+	})
+})
+
+var _ = Describe("func UnmarshalTime()", func() {
```
Comment (koden-km) [inline envelope_test.go:95]:
```suggestion
var _ = Describe("func UnmarshalEnvelopeTime()", func() {
```

**Review by jmalloc — CHANGES_REQUESTED** [review]

**Review by koden-km — CHANGES_REQUESTED** [review]

**Review by danilvpetrov — COMMENTED** [review]
Leaving a comment so I can commit my own suggestion 😄

**Review by jmalloc — APPROVED** [review]

**Comment by danilvpetrov** [comment]
Sorry @koden-km I am merging this PR in, LMK if you have some comments later on.

**Pull request merged**